<h3><b>Introduction to Data Analytics - Exercise set 2 - pandas-module</b></h3>

Put all your exercises (Jupyter Notebooks or Python-files) in your course Git-project.
Use either code comments or Jupyter Notebook markdown (text) to document which exercise you are doing and what a certain code section does! 
You can return all of these exercises in a single Jupyter Notebook, if you wish.

<b>The datasets for these exercises have been collected from kaggle.com<br />
(a service providing different datasets for practice)</b>

<img src="http://srv.plab.fi/~tuomasv/data_analytics_2023_images/exercise_set_2/es2_1.png" />

<h4><b>1. import pandas and read the csv-file found in Moodle (loans.csv). Use Python coding with pandas to answer the questions.</b></h4>


In [78]:
# you can import numpy and pandas here
# and also load the needed csv-file
# I imported dtale module for myslef 
# it helped me to do the set
import numpy as np
import pandas as pd
import dtale

df = pd.read_csv("loans.csv")
df

Loan ID  \
0       14dd8831-6af5-400b-83ec-68e61888a048   
1       4771cc26-131a-45db-b5aa-537ea4ba5342   
2       4eed4e6a-aa2f-4c91-8651-ce984ee8fb26   
3       77598f7b-32e7-4e3b-a6e5-06ba0d98fe8a   
4       d4062e70-befa-4995-8643-a0de73938182   
...                                      ...   
100509                                   NaN   
100510                                   NaN   
100511                                   NaN   
100512                                   NaN   
100513                                   NaN   

                                 Customer ID Loan Status  Current Loan Amount  \
0       981165ec-3274-42f5-a3b4-d104041a9ca9  Fully Paid             445412.0   
1       2de017a3-2e01-49cb-a581-08169e83be29  Fully Paid             262328.0   
2       5efb2b2b-bf11-4dfd-a572-3761a2694725  Fully Paid           99999999.0   
3       e777faab-98ae-45af-9a86-7ce5b33b1011  Fully Paid             347666.0   
4       81536ad9-5ccf-4eb8-befb-47a4d608658e  Fully Paid             176220.0   
...                                      ...         ...                  ...   
100509                                   NaN         NaN                  NaN   
100510                                   NaN         NaN                  NaN   
100511                                   NaN         NaN                  NaN   
100512                                   NaN         NaN                  NaN   
100513                                   NaN         NaN                  NaN   

              Term  Credit Score  Annual Income Years in current job  \
0       Short Term         709.0      1167493.0              8 years   
1       Short Term           NaN            NaN            10+ years   
2       Short Term         741.0      2231892.0              8 years   
3        Long Term         721.0       806949.0              3 years   
4       Short Term           NaN            NaN              5 years   
...            ...           ...            ...                  ...   
100509         NaN           NaN            NaN                  NaN   
100510         NaN           NaN            NaN                  NaN   
100511         NaN           NaN            NaN                  NaN   
100512         NaN           NaN            NaN                  NaN   
100513         NaN           NaN            NaN                  NaN   

       Home Ownership             Purpose  Monthly Debt  \
0       Home Mortgage   Home Improvements       5214.74   
1       Home Mortgage  Debt Consolidation      33295.98   
2            Own Home  Debt Consolidation      29200.53   
3            Own Home  Debt Consolidation       8741.90   
4                Rent  Debt Consolidation      20639.70   
...               ...                 ...           ...   
100509            NaN                 NaN           NaN   
100510            NaN                 NaN           NaN   
100511            NaN                 NaN           NaN   
100512            NaN                 NaN           NaN   
100513            NaN                 NaN           NaN   

        Years of Credit History  Months since last delinquent  \
0                          17.2                           NaN   
1                          21.1                           8.0   
2                          14.9                          29.0   
3                          12.0                           NaN   
4                           6.1                           NaN   
...                         ...                           ...   
100509                      NaN                           NaN   
100510                      NaN                           NaN   
100511                      NaN                           NaN   
100512                      NaN                           NaN   
100513                      NaN                           NaN   

        Number of Open Accounts  Number of Credit Problems  \
0                           6.0                        1.0   
1       

<li>Remove the Customer ID –column from data</li>
<li>Print the head of the data</li>

In [79]:
# remove the column and print out the head of the DataFrame
df = df.drop(['Customer ID'], axis=1)
df

Loan ID Loan Status  Current Loan Amount  \
0       14dd8831-6af5-400b-83ec-68e61888a048  Fully Paid             445412.0   
1       4771cc26-131a-45db-b5aa-537ea4ba5342  Fully Paid             262328.0   
2       4eed4e6a-aa2f-4c91-8651-ce984ee8fb26  Fully Paid           99999999.0   
3       77598f7b-32e7-4e3b-a6e5-06ba0d98fe8a  Fully Paid             347666.0   
4       d4062e70-befa-4995-8643-a0de73938182  Fully Paid             176220.0   
...                                      ...         ...                  ...   
100509                                   NaN         NaN                  NaN   
100510                                   NaN         NaN                  NaN   
100511                                   NaN         NaN                  NaN   
100512                                   NaN         NaN                  NaN   
100513                                   NaN         NaN                  NaN   

              Term  Credit Score  Annual Income Years in current job  \
0       Short Term         709.0      1167493.0              8 years   
1       Short Term           NaN            NaN            10+ years   
2       Short Term         741.0      2231892.0              8 years   
3        Long Term         721.0       806949.0              3 years   
4       Short Term           NaN            NaN              5 years   
...            ...           ...            ...                  ...   
100509         NaN           NaN            NaN                  NaN   
100510         NaN           NaN            NaN                  NaN   
100511         NaN           NaN            NaN                  NaN   
100512         NaN           NaN            NaN                  NaN   
100513         NaN           NaN            NaN                  NaN   

       Home Ownership             Purpose  Monthly Debt  \
0       Home Mortgage   Home Improvements       5214.74   
1       Home Mortgage  Debt Consolidation      33295.98   
2            Own Home  Debt Consolidation      29200.53   
3            Own Home  Debt Consolidation       8741.90   
4                Rent  Debt Consolidation      20639.70   
...               ...                 ...           ...   
100509            NaN                 NaN           NaN   
100510            NaN                 NaN           NaN   
100511            NaN                 NaN           NaN   
100512            NaN                 NaN           NaN   
100513            NaN                 NaN           NaN   

        Years of Credit History  Months since last delinquent  \
0                          17.2                           NaN   
1                          21.1                           8.0   
2                          14.9                          29.0   
3                          12.0                           NaN   
4                           6.1                           NaN   
...                         ...                           ...   
100509                      NaN                           NaN   
100510                      NaN                           NaN   
100511                      NaN                           NaN   
100512                      NaN                           NaN   
100513                      NaN                           NaN   

        Number of Open Accounts  Number of Credit Problems  \
0                           6.0                        1.0   
1                          35.0                        0.0   
2                          18.0                        1.0   
3                           9.0                        0.0   
4                          15.0                        0.0   
...                         ...                        ...   
100509                      NaN                        NaN   
100510                      NaN                        NaN   
100511                      NaN                        NaN   
100512                      NaN                        NaN   
100513                     

<li>Remove rows from the data that have <b style="color: red;">a too large of a loan</b> (Current Loan Amount should be less than 99999999)</li>

In [80]:
# remove loans that are too large
# for this we need to download Spicy module
from scipy import stats
#df = df[(np.abs(stats.zscore(df)) < 3).all(axis=1)]

# extra task: use imputation to use average income as the value instead of NaN
df['Annual Income'].fillna((df['Annual Income'].mean()), inplace=True)

<ul>
    <li>Remove rows that have the annual income as NaN (not a number)</li>
    <ul>
        <li><b>Extra task:</b> use imputation to use average income as the value instead of NaN</li>
    </ul>
</ul>

In [81]:
# remove rows if annual income is NaN
df = df[df['Annual Income'].notnull()]

<li>Get the average Current Loan Amount</li>

In [82]:
# get the average current loan
mean = df['Current Loan Amount'].mean()

<li>Get the highest and lowest Annual Income in the dataset</li>

In [83]:
# get highest and lowest annual income in dataset
min = df['Annual Income'].min()
max = df['Annual Income'].max()

<li>Get the Home Ownership value of the <b>Loan ID = bbf87a87-22cd-4d10-bd9b-7a9cc1b6e59d</b></li>

In [84]:
# get the needed value with the Loan ID
df = df.rename(columns={"Home Ownership": "Home_Ownership"})
Loan_id = "bbf87a87-22cd-4d10-bd9b-7a9cc1b6e59d"
Home_Ownership = df[df['Loan ID'] == Loan_id]
Home_Ownership[["Home_Ownership"]]

Home_Ownership
491       Own Home

<li>Get the Actual Annual Income of the loan with the <b>ID = 76fa89b9-e6a8-49af-afa1-8151315aba8e</b></li>
<br><b>Note:</b> The Actual Annual Income should be :
<b style="color: green">Annual Income – 12 * Monthly Debt</b><br><br>
<b>Tip:</b> Create a new field with the above instruction to the DataFrame.


In [85]:
# create a new field for Actual Annual income (annual income - 12 * monthly debt)
# get the Actual Annual income for the needed 

df["Actual Annual Income"] = df["Monthly Debt"] * 12
df.loc[df ["Loan ID"] == "76f a89b9-e6a8-49af-afal-8151315abage", "Actual Annual Income"]

Series([], Name: Actual Annual Income, dtype: float64)

<li>Get the Loan ID of the loan with the smallest Actual Annual Income </li>

In [86]:
# get loan ID of lowest actual annual income
lowest_income = df.loc[df['Actual Annual Income'].idxmin()]['Loan ID']
lowest_income

'da795dc7-9960-49d9-9ef4-94a28a51f168'

<li>How many loans are "Long term"?</li>

In [87]:
# produce code that shows how many loans are long term
loans = df[df["Term"] == "Long Term"]
loans

Loan ID  Loan Status  Current Loan Amount  \
3      77598f7b-32e7-4e3b-a6e5-06ba0d98fe8a   Fully Paid             347666.0   
7      db0dc6e1-77ee-4826-acca-772f9039e1c7  Charged Off             648714.0   
15     7cbaa3fa-16fd-4343-9bcb-e90b34a1072f   Fully Paid             449020.0   
16     c9a16a9d-8801-4430-b445-dbf9cf845e31  Charged Off             653004.0   
17     24e8c8bd-d10b-4dac-8b81-1da470ff5ecb   Fully Paid             666204.0   
...                                     ...          ...                  ...   
99971  fa1a1ef5-3697-469c-a842-538b8dcfb63a   Fully Paid           99999999.0   
99972  66a1c384-c294-4612-9705-d79aa95cf6bd   Fully Paid             328350.0   
99973  15c4eb40-173f-4035-a983-285c5b5aac1d   Fully Paid             522610.0   
99978  94eaa3d9-2b9f-467b-855f-9e8f5eae57a4   Fully Paid             431860.0   
99982  81389d2c-963c-40d4-a8eb-8329a3efadad   Fully Paid             484968.0   

            Term  Credit Score  Annual Income Years in current job  \
3      Long Term         721.0   8.069490e+05              3 years   
7      Long Term           NaN   1.378277e+06             < 1 year   
15     Long Term           NaN   1.378277e+06              9 years   
16     Long Term           NaN   1.378277e+06              7 years   
17     Long Term         723.0   1.821967e+06            10+ years   
...          ...           ...            ...                  ...   
99971  Long Term         616.0   1.530564e+06              8 years   
99972  Long Term         704.0   2.041740e+06              5 years   
99973  Long Term         725.0   1.026209e+06            10+ years   
99978  Long Term           NaN   1.378277e+06            10+ years   
99982  Long Term           NaN   1.378277e+06            10+ years   

      Home_Ownership             Purpose  Monthly Debt  \
3           Own Home  Debt Consolidation       8741.90   
7      Home Mortgage           Buy House      14806.13   
15          Own Home  Debt Consolidation      18904.81   
16     Home Mortgage  Debt Consolidation      14537.09   
17     Home Mortgage  Debt Consolidation      17612.24   
...              ...                 ...           ...   
99971           Rent       Medical Bills      13264.85   
99972           Rent               other      32684.94   
99973  Home Mortgage  Debt Consolidation      15307.54   
99978  Home Mortgage  Debt Consolidation       8465.07   
99982  Home Mortgage  Debt Consolidation      20941.80   

       Years of Credit History  Months since last delinquent  \
3                         12.0                           NaN   
7                          8.2                           8.0   
15                        19.4                           NaN   
16                        20.5                           NaN   
17                        22.0                          34.0   
...                        ...                           ...   
99971                     28.0                           3.0   
99972                     16.7                           NaN   
99973                     29.4                           NaN   
99978                     18.5                          25.0   
99982                     23.7                           NaN   

       Number of Open Accounts  Number of Credit Problems  \
3                          9.0                        0.0   
7                         15.0                        0.0   
15                         8.0                        0.0   
16                         9.0                        0.0   
17                        15.0                        0.0   
...                        ...                        ...   
99971                     10.0                        0.0   
99972                      9.0                        0.0   
99973                     11.0                        0.0   
99978                      5.0                        0.0   
99982                     15.0                        0.0   

       Current Credit B

<li>How many loaners have more than 1 bankruptcy?</li>

In [88]:
# produce code that shows how many loaners have more than 1 bankruptcy
loans = df[df["Bankruptcies"] > 1]
loans

Loan ID  Loan Status  Current Loan Amount  \
180    32200a7e-c65e-4e5e-9878-1161d256bcfe   Fully Paid             234806.0   
369    2fc4e87c-eb83-4dd2-9d7e-b09b85c2e812   Fully Paid              33154.0   
777    6ba36bae-c42d-4aeb-889b-cda78bf317b5   Fully Paid             172436.0   
948    2ca16db4-8279-4bc1-bb02-180503c8d127  Charged Off             298760.0   
967    32200a7e-c65e-4e5e-9878-1161d256bcfe   Fully Paid           99999999.0   
...                                     ...          ...                  ...   
99542  3fb3d12f-cf5d-4d8a-901a-0f898464b60f   Fully Paid             220682.0   
99629  69d7fe10-4ad8-440a-b9a5-6dd764bf4c20   Fully Paid           99999999.0   
99894  67c58dcc-3a3e-4c0a-ac3f-f453f6e60a4d  Charged Off             403898.0   
99895  a8cf322f-c968-4c06-aa1d-50fcef54e760  Charged Off             120384.0   
99954  e367cc19-d452-481a-ac34-c5bdeba2cb6e  Charged Off             316426.0   

             Term  Credit Score  Annual Income Years in current job  \
180     Long Term         689.0   8.667990e+05              9 years   
369    Short Term         713.0   5.727930e+05            10+ years   
777    Short Term         740.0   1.340222e+06              7 years   
948    Short Term           NaN   1.378277e+06            10+ years   
967     Long Term         689.0   8.667990e+05              9 years   
...           ...           ...            ...                  ...   
99542  Short Term         746.0   8.481220e+05                  NaN   
99629  Short Term         725.0   1.133616e+06            10+ years   
99894   Long Term         689.0   8.745320e+05              3 years   
99895  Short Term         723.0   4.084620e+05                  NaN   
99954   Long Term         659.0   1.613062e+06               1 year   

      Home_Ownership             Purpose  Monthly Debt  \
180    Home Mortgage  Debt Consolidation       3676.69   
369             Rent            vacation      13412.86   
777         Own Home  Debt Consolidation      14965.92   
948             Rent  Debt Consolidation      11952.14   
967    Home Mortgage  Debt Consolidation       3676.69   
...              ...                 ...           ...   
99542  Home Mortgage  Debt Consolidation      14135.43   
99629  Home Mortgage  Debt Consolidation      21633.21   
99894       Own Home  Debt Consolidation      20697.27   
99895           Rent  Debt Consolidation       2617.63   
99954           Rent  Debt Consolidation      15189.74   

       Years of Credit History  Months since last delinquent  \
180                       14.1                           7.0   
369                       11.8                           NaN   
777                       19.8                           NaN   
948                       17.3                          63.0   
967                       14.1                           7.0   
...                        ...                           ...   
99542                     16.2                          68.0   
99629                     14.9                          23.0   
99894                     12.7                          74.0   
99895                     22.0                          35.0   
99954                     21.5                           NaN   

       Number of Open Accounts  Number of Credit Problems  \
180                        4.0                        2.0   
369                       10.0                        2.0   
777                        6.0                        7.0   
948                        9.0                        6.0   
967                        4.0                        2.0   
...                        ...                        ...   
99542                      9.0                        3.0   
99629                     16.0                        4.0   
99894                      9.0                        2.0   
99895                      4.0                        3.0   
99954                     25.0                        2.0   

       Curr

<li>How many Short Term loans are for Home Improvements?</li>

In [96]:
# produce code that shows how many of the Short Term loans are for Home Improvements
df = df.dropna()

loans = df[(df["Purpose"] == "Home Improvements") & (df["Term"] == "Short Term")]
loans

Loan ID  Loan Status  Current Loan Amount  \
93     9270a0f9-c873-4cff-b8d6-63b5377a0340   Fully Paid             156772.0   
140    26732d68-bf40-4041-a7fc-32893ba51a5c  Charged Off             232716.0   
173    0b074d25-97a5-4a2f-9df1-a0efe04f597c   Fully Paid             716958.0   
328    a145decf-d736-44dd-9da0-44d87cfeb59e   Fully Paid           99999999.0   
353    9e3fe95c-0305-4314-bd81-aee93ea6d07e  Charged Off             133936.0   
...                                     ...          ...                  ...   
99691  ce112ac6-1c24-460e-9bd1-b6e61b52ab88   Fully Paid           99999999.0   
99786  e7ec7678-e681-46fb-95b6-431731fef2bc   Fully Paid             212828.0   
99905  b98dda11-6d48-4253-a670-0f02c0da570b   Fully Paid             679778.0   
99917  822750aa-26d4-43b7-87f7-dda3ebfb7e2d   Fully Paid             431332.0   
99939  8e931dbb-162c-432a-8be8-45bf95a830e1   Fully Paid             222530.0   

             Term  Credit Score  Annual Income Years in current job  \
93     Short Term         748.0      1411966.0              6 years   
140    Short Term         637.0      1049427.0              5 years   
173    Short Term         718.0      1934960.0              9 years   
328    Short Term         716.0      2848575.0             < 1 year   
353    Short Term         639.0       347035.0               1 year   
...           ...           ...            ...                  ...   
99691  Short Term         744.0       781261.0             < 1 year   
99786  Short Term         664.0      1028907.0            10+ years   
99905  Short Term         704.0      3302029.0             < 1 year   
99917  Short Term         719.0       931285.0              2 years   
99939  Short Term         751.0      1825767.0              6 years   

      Home_Ownership            Purpose  Monthly Debt  \
93     Home Mortgage  Home Improvements      22591.38   
140             Rent  Home Improvements      12942.99   
173    Home Mortgage  Home Improvements      31765.72   
328    Home Mortgage  Home Improvements      23263.41   
353    Home Mortgage  Home Improvements       6969.39   
...              ...                ...           ...   
99691  Home Mortgage  Home Improvements      11718.82   
99786  Home Mortgage  Home Improvements      12946.98   
99905  Home Mortgage  Home Improvements      17830.93   
99917  Home Mortgage  Home Improvements      14667.81   
99939  Home Mortgage  Home Improvements      11152.24   

       Years of Credit History  Months since last delinquent  \
93                        18.0                          73.0   
140                        9.5                          61.0   
173                       10.0                          24.0   
328                       31.5                          10.0   
353                       15.4                          22.0   
...                        ...                           ...   
99691                     14.9                          76.0   
99786                     13.1                          39.0   
99905                     17.6                          22.0   
99917                     14.5                          41.0   
99939                     26.0                          73.0   

       Number of Open Accounts  Number of Credit Problems  \
93                        10.0                        0.0   
140                       20.0                        0.0   
173                        9.0                        0.0   
328                       14.0                        0.0   
353                       10.0                        0.0   
...                        ...                        ...   
99691                     14.0                        0.0   
99786                     13.0                        0.0   
99905                      6.0                        0.0   
99917                     11.0                        0.0   
99939                     21.0                        0.0   

       Current Credit B

<li>How many unique loan purposes are there?</li>

In [93]:
# produce code that shows how many unique loan purposes there are
unique_purpose_numbers = df['Purpose'].nunique()
unique_purpose_numbers

16

<li>What are the 3 most common loan purposes?</li>

In [101]:
# produce code that shows the 3 most common loan purposes
purposes = df['Purpose'].value_counts()
purposes.head(3)

Debt Consolidation    28202
Home Improvements      2409
other                  2313
Name: Purpose, dtype: int64

<li>Is there a correlation between <b>Annual Income</b> and <b>Number of Open Accounts</b> or is there a correlation between <b>Number of Credit Problems</b> and <b>Bankruptcies</b>?</li>

In [103]:
# produce code that shows which correlation statement is true
dtale.show(df).open_browser()

# 2. Download the purchases.csv from Moodle, and do the following observations:

<li>What was the total price sum of the Purchase Order Number 018H2015? (14 rows in total)</li>

In [ ]:
# produce the code that gets the needed total price sum
#df = pd.read_csv("purchases.csv")

#sum = df[df['Purchase Order Number' == "018H2015"].sum()
#sum

<li>What is the name and description of the purchased item with the Purchase Order Number 3176273?</li>

In [ ]:
# produce the code that gets the needed name and description

<li>How many occasions (rows) of purchase data happened during the year 2013?</li>

In [ ]:
# produce the code that filters the needed data for year 2013

<ul><li>What are the 5 most common Departments in the data?</li>
<ul><li><b>Extra task: </b>What are 3 Departments using most money in the data?</li></ul>
</ul>

In [ ]:
# produce the code that gtest the needed most common departments

<li>Sort the data by Department Name</li>

In [ ]:
# sort the data by department name

<ul>
<li><b>Small extra tasks for extra points</b></li>
<ul>
<li>How many purchases in the data were IT Goods and had the total price more than 50000 dollars?</li>
<li>How many of the purchases have anything to do with IT? (IT Goods, IT Services, IT Telecommunications)</li>

</ul>
</ul>

In [ ]:
# produce the code for the extra tasks if you wish complete them

<img src="http://srv.plab.fi/~tuomasv/data_analytics_2023_images/exercise_set_2/es2_2.png" />

<h4><b>3. Download the data_salaries_india.csv from Moodle, and consider the following questions of the data. Use any means in pandas (or even NumPy) you wish to explain your answers.</b></h4>


<ul>
<li>Before we can do anything with the salaries, we have to convert them into something more usable</li>
<ul>
<li>Note: the salaries can be yearly, monthly or hourly salaries</li>
<ul>
<li>We don't also need the Indian rupee –sign (₹)</li>
</ul>
<li>You can use the template in Moodle to help you out with this (Salary filtering, pandas exercise 3)</li>
</ul>
</ul>

In [ ]:
# convert the salaries as needed

<ul>
<li>What are the most common values in different fields (Job Titles, Companies, Location)? <b>Based on the distribution, is the data balanced or not?</b></li>
<ul>
<li><b>Extra task:</b> there seem to be some Job Titles that are almost the same, like "Machine Learning Data Associate" and "Machine Learning Associate", combine these into something common</li>
</ul>
</ul>

In [ ]:
# produce the code that shows the most common values in different fields
# and answer the question about the distribution

<li>Are there any outliers in the data that might affect the averages negatively (certain salaries)? Manage the outliers as you best see fit (either remove them or leave them, based on your analysis) </li>

In [ ]:
# produce the code that shows the potential outliers and take care of them as you see fit

<ul>
<li>If we want to correlate upon categories (ordinal or binary), we need to use factorize(). <b>Factorize the Role-column, and add the new column to the DataFrame.</b></li>
<ul>
<li><b>Note:</b> using <b>factorize()</b> for nominal categories (Job Title, Location, Company Name) doesn't work well, because the numbers do not have any numeric magnitude. In other words, these categories don't measure anything, they just group data, so numerical comparison / correlation doesn't mean anything statistically.</li>
</ul>
</ul>

In [ ]:
# example:
# factorize Role level into numbers
# label1, unique1 = pd.factorize(df['Role'], sort=False)
# df['ManagerRole'] = label1

# produce the needed factorization for Role-column

<ul>
<li><b>Finally, check out the correlations. Does anything correlate with anything? Can we make any assumptions?</b></li>
<ul>
<li>Tip: When correlating against binary variables, sometimes the Spearman correlation might be more sensitive, in pandas:</li>
</ul>
</ul>

In [ ]:
# Remember: df.corr(method="spearman")

<b>Smaller extra tasks</b><br><br>
<ul>
<li>Did you get any ideas how this data could be improved? Do we need some particular new data or some other operations on the data? Should we filter something out based on some other column?<br><br><b>Provide arguments for your answers in code comments.</b><br><br> </li>
<ul>
<li>Note: There are many good possible answers here!</li>
<li>Tip: How about replacing the "Salaries Reported" column with actual rows based on that number? Try doing this with the data!</li>
<li>Remember: This data only represents data engineering salaries based on selected Indian cities.
The world is a vast place :)
</li>
</ul>
</ul>

In [ ]:
# code for the smaller extra tasks


<h2><b>Advanced extra tasks for extra points (varying challenges, some require Googling):</b></h2>

<ul>
<li><b>1. Data merge is a useful tool when you have multiple files of data that have the exact same structure.</b><br><br>Download the two csv-files from Moodle (videogames1.csv, videogames2.csv), and combine them into one Data Frame. Lastly, save the Data Frame into a new csv-file => combined.csv.</li>
</ul>

In [ ]:
# code for the advanced extra task

<ul>
<li><b>2. Functions and lambdas allow us to extend the operations we wish to do to columns and rows in pandas. </b><br><br>For example, the built-in functions may not be enough in all cases. Use the data of exercise 1 (loans.csv), and create a new column called "Income Group" that holds a text value. <br><br>Create either a function or a lambda, that determines the True or False –value based on the Annual Income –column. Use the following table to create the values:</li>
</ul>

<img src="http://srv.plab.fi/~tuomasv/data_analytics_2023_images/exercise_set_2/es2_3.png"/>

<li>After creating the function/lambda, you can use it by using pandas' .apply() –function.<br><br></li>

<li>Finally, get the amount of rows grouped by each of the new Income Group –field values and print them out.</li> 


In [ ]:
# code for the advanced extra task

<ul>
<li><b>3.	Normalization allows us to convert the values of numeric columns to be between 0 and 1.</b> This is helpful when two different numbers seem to follow the same trend, but have completely different value ranges. For example, gold and silver prices tend to follow similar patterns, but their market worth is quite different. By using normalization, we can compare these trends more easily. <br>

Get historical prices of both gold and silver, and compare them without and with normalization. You can plot the prices by using df.plot() –function. Check the dataset in list in Moodle for some alternatives for gold and silver prices.
</li>
</ul>
<img src="http://srv.plab.fi/~tuomasv/data_analytics_2023_images/exercise_set_2/es2_4.png" />

<ul>
<li><b>4.	Create an account to kaggle.com, and find any dataset that interests you. </b><br>

There's a list of possibly interesting datasets listed in Moodle as well.
<br><br>
<b>Try to find interesting features in data, in particular:</b>
</li>
<ul>
<li>Clean up data first (rows with too many NaN –values), values that are way too big or small, insignificant columns etc.)</li>
<li>You can create new columns as well if it seems suitable! (either by using functions or other means)</li>
<li>Interesting correlations (.corr() –function) and other interesting features in the data. Is something surprising in the data?</li>
<li><b>Note: </b>There are many ways on how to approach this exercise.</li>
</ul>
</ul>

<img src="http://srv.plab.fi/~tuomasv/data_analytics_2023_images/exercise_set_2/es2_5.png" />

In [ ]:
# do any Kaggle.com -related extra tasks in their own Jupyter notebooks for easier coding